### PR TITLE
multithreadedhttpconnectionmanager clean up

### PR DESCRIPTION
**Pull request by marianp123**
see new MultiThreadedHttpConnectionManagerCleanUp.java

### CODE REVIEW

**Comment by mjiderhamn** [comment]
Thanks for the Pull Request!

File: classloader-leak-prevention/src/test/java/se/jiderhamn/classloader/leak/prevention/MultiThreadedHttpConnectionManagerCleanUpTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import se.jiderhamn.classloader.leak.Leaks;
+import se.jiderhamn.classloader.leak.prevention.cleanup.MultiThreadedHttpConnectionManagerCleanUp;
+
+/**
+ * Test case for leaks caused by com.sun.jersey.client.apache.ApacheHttpClient failing to close org.apache.commons.httpclient.MultiThreadedHttpConnectionManager
+ * 
+ * @author Marian Petrik
+ */
+@RunWith(JUnitClassloaderRunner.class)
+public class MultiThreadedHttpConnectionManagerCleanUpTest extends PreventionsTestBase<MultiThreadedHttpConnectionManagerCleanUp>{
```
Comment (mjiderhamn) [inline classloader-leak-prevention/src/test/java/se/jiderhamn/classloader/leak/prevention/MultiThreadedHttpConnectionManagerCleanUpTest.java]: Any reason this test class cannot be simplified to a subclass of `ClassLoaderPreMortemCleanUpTestBase` (see for example [DriverManagerCleanUpTest](https://github.com/mjiderhamn/classloader-leak-prevention/blob/master/classloader-leak-prevention/src/test/java/se/jiderhamn/classloader/leak/prevention/cleaup/DriverManagerCleanUpTest.java))?

File: classloader-leak-prevention/src/test/java/se/jiderhamn/classloader/leak/prevention/MultiThreadedHttpConnectionManagerCleanUpTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import com.sun.jersey.api.client.WebResource;
+import com.sun.jersey.api.client.config.DefaultClientConfig;
+import com.sun.jersey.client.apache.ApacheHttpClient;
+
+import se.jiderhamn.classloader.leak.JUnitClassloaderRunner;
+import se.jiderhamn.classloader.leak.Leaks;
+import se.jiderhamn.classloader.leak.prevention.cleanup.MultiThreadedHttpConnectionManagerCleanUp;
+
+/**
+ * Test case for leaks caused by com.sun.jersey.client.apache.ApacheHttpClient failing to close org.apache.commons.httpclient.MultiThreadedHttpConnectionManager
```
Comment (mjiderhamn) [inline classloader-leak-prevention/src/test/java/se/jiderhamn/classloader/leak/prevention/MultiThreadedHttpConnectionManagerCleanUpTest.java]: (For consistency, consider using `{@link ...}` to reference the classes)

File: classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/ClassLoaderLeakPreventorListener.java [hunk 3]
```diff
@@ -237,6 +237,8 @@ public ClassLoaderLeakPreventorListener() {
       classLoaderLeakPreventorFactory.addPreInitiator(new OracleJdbcThreadInitiator());
 
     classLoaderLeakPreventorFactory.addCleanUp(new BeanIntrospectorCleanUp());
+    
+    classLoaderLeakPreventorFactory.addCleanUp(new MultiThreadedHttpConnectionManagerCleanUp());
```
Comment (mjiderhamn) [inline classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/ClassLoaderLeakPreventorListener.java]: Would you mind moving this to line 252, to maintain alphabetical order?

File: classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/MultiThreadedHttpConnectionManagerCleanUp.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+public class MultiThreadedHttpConnectionManagerCleanUp implements ClassLoaderPreMortemCleanUp {
+	
+  @Override
+  public void cleanUp(ClassLoaderLeakPreventor preventor) {
+	    final Class<?> connManager = preventor.findClass("org.apache.commons.httpclient.MultiThreadedHttpConnectionManager");
+	    if(connManager != null) {
+	      try {
+	    	  connManager.getMethod("shutdownAll").invoke(null);
+	      }
+	      catch (Exception ex) {
```
Comment (mjiderhamn) [inline classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/MultiThreadedHttpConnectionManagerCleanUp.java]: To be consistent, this should catch `Throwable` and log an error or warning.

File: classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/MultiThreadedHttpConnectionManagerCleanUp.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ * @author Marian Petrik
+ */
+public class MultiThreadedHttpConnectionManagerCleanUp implements ClassLoaderPreMortemCleanUp {
+	
+  @Override
+  public void cleanUp(ClassLoaderLeakPreventor preventor) {
+	    final Class<?> connManager = preventor.findClass("org.apache.commons.httpclient.MultiThreadedHttpConnectionManager");
+	    if(connManager != null) {
+	      try {
+	    	  connManager.getMethod("shutdownAll").invoke(null);
```
Comment (mjiderhamn) [inline classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/MultiThreadedHttpConnectionManagerCleanUp.java]: I though `ClassLoaderLeakPreventor` already had an `invokeStaticMethod(Class clazz, String name)`, but this seems like a good time to create one :-)

File: classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/MultiThreadedHttpConnectionManagerCleanUp.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package se.jiderhamn.classloader.leak.prevention.cleanup;
+
+import se.jiderhamn.classloader.leak.prevention.ClassLoaderLeakPreventor;
+import se.jiderhamn.classloader.leak.prevention.ClassLoaderPreMortemCleanUp;
+
+/**
+ * Invokes static method org.apache.commons.httpclient.MultiThreadedHttpConnectionManager.shutdownAll() to close connections left out by com.sun.jersey.client.apache.ApacheHttpClient.
```
Comment (mjiderhamn) [inline classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/MultiThreadedHttpConnectionManagerCleanUp.java]: Is there some documentation regarding this that would be relevant to link here?

**Comment by marianp123** [comment]
Hi,
I have fixed all of the comments except for exception logging, because I expect the exception to be thrown with a non-existing url. we do not need to make a call to a real url, only to set up relevant classes for the leak to appear.

**Comment by mjiderhamn** [comment]
> I have fixed all of the comments except for exception logging, because I expect the exception to be thrown with a non-existing url

Thanks, however my comment on exception logging was in the `MultiThreadedHttpConnectionManagerCleanUp` implementation, not in the test case.

File: classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/MultiThreadedHttpConnectionManagerCleanUp.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ * Invokes static method org.apache.commons.httpclient.MultiThreadedHttpConnectionManager.shutdownAll() to close connections left out by com.sun.jersey.client.apache.ApacheHttpClient.  
+ * 
+ * @author Marian Petrik
+ */
+public class MultiThreadedHttpConnectionManagerCleanUp implements ClassLoaderPreMortemCleanUp {
+	
+  @Override
+  public void cleanUp(ClassLoaderLeakPreventor preventor) {
+	    final Class<?> connManager = preventor.findClass("org.apache.commons.httpclient.MultiThreadedHttpConnectionManager");
+	    if(connManager != null) {
```
Comment (mjiderhamn) [inline classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/MultiThreadedHttpConnectionManagerCleanUp.java]: Do you want to check that the class is loaded by the protected `ClassLoader` (`preventor.isLoadedByClassLoader(connManager)`)? Or what would happen if Jersey was supplied on the application server level?

Comment (marianp123) [inline classloader-leak-prevention/src/main/java/se/jiderhamn/classloader/leak/prevention/cleanup/MultiThreadedHttpConnectionManagerCleanUp.java]: I found the bug in an application running on JBoss EAP with Jersey dependencies provided by the application server. I tried overriding the dependencies to compile scope newest version, didn't help. I am not going to do anything more about this pull request. Thanks.

**Comment by mjiderhamn** [comment]
@marianp123 Did the existing thread stopping feature of this library not take care of this case?

**Pull request merged**